deps/v8/src/builtins/builtins-trace.cc

```diff
@@ -0,0 +1,701 @@
+// Copyright 2018 the V8 project authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "src/builtins/builtins-utils.h"
+#include "src/builtins/builtins.h"
+#include "src/counters.h"
+#include "src/api.h"
+#include "src/objects-inl.h"
+
+namespace v8 {
+namespace internal {
+
+namespace {
+
+using v8::tracing::TracedValue;
+
+enum Result { NODATA, SUCCESS, FAILED };
+
+class MaybeUtf8 {
+ public:
+  MaybeUtf8(Handle<String> string) {
+    string = String::Flatten(string);
+    one_byte_ = string->IsOneByteRepresentation();
+    if (one_byte_) {
+      data_ = string->length() == 0 ? "" :
+          reinterpret_cast<const char*>(Handle<SeqOneByteString>::cast(string)
+              ->GetChars());
+    } else {
+      data_ = nullptr;
+      Local<v8::String> local = Utils::ToLocal(string);
```

Comment:
I would not do this round trip through the V8 API. Also one-byte strings can include Latin1 characters, which are not UTF8 encoding?